python/python/bystro/covariance/_base_covariance.py

```diff
@@ -0,0 +1,549 @@
+###########################################
+###########################################
+###                                     ###
+###                                     ###
+###   Methods for covariance matrices   ###
+###                                     ###
+###                                     ###
+###########################################
+###########################################
+###########################################
+
+
+def _get_precision(covariance):
+    """
+    Gets the precision matrix defined as the inverse of the covariance
+
+    Parameters
+    ----------
+    covariance : np.array-like(p,p)
+        The covariance matrix
+
+    Returns
+    -------
+    precision : np.array-like(sum(p),sum(p))
+        The inverse of the covariance matrix
+    """
+    precision = la.inv(covariance)
+    return precision
+
+
+def _get_stable_rank(covariance):
+    """
+    Returns the stable rank defined as
+    ||A||_F^2/||A||^2
+
+    Parameters
+    ----------
+    covariance : np.array-like(p,p)
+        The covariance matrix
+
+    Returns
+    -------
+    srank : float
+        The stable rank
+    """
+    singular_values = la.svd(covariance, compute_uv=False)
+    srank = np.sum(singular_values) / singular_values[0]
+    return srank
+
+
+def _predict(covariance, Xobs, idxs):
+    """
+    Predicts missing data using observed data.
+
+    Parameters
+    ----------
+    covariance : np.array-like(p,p)
+        The covariance matrix
+
+    Xobs : np array-like,(N_samples,\sum idxs)
+        The observed data
+
+    idxs: np.array-like,(sum(p),)
+        The observation locations
+
+    Returns
+    -------
+    preds : np.array-like,(N_samples,p-\sum idxs)
+        The predicted values
+    """
+    covariance_sub = covariance[idxs == 1]
+    covariance_22 = covariance_sub[:, idxs == 1]
+    covariance_21 = covariance_sub[:, idxs == 0]
+
+    covariance_nonsub = covariance[idxs == 0]
+    covariance_11 = covariance_nonsub[:, idxs == 0]
+
+    beta_bar = la.solve(covariance_22, covariance_21)
+    Second_part = np.dot(covariance_21.T, beta_bar)
+
+    covariance_11 - Second_part
+
+    preds = np.dot(Xobs[:, idxs == 1], beta_bar)
+    return preds
+
+
+#########################################
+#########################################
+##                                     ##
+##  Gaussian Likelihood-based methods  ##
+##                                     ##
+#########################################
+#########################################
+
+
+def _conditional_score(covariance, X, idxs, weights=None):
+    """
+    Returns the predictive log-likelihood of a subset of data.
+
+    mean(log p(X[idx==1]|X[idx==0],covariance))
+
+    Parameters
+    ----------
+    covariance : np.array-like(p,p)
+        The covariance matrix
+
+    X : np.array-like,(N,sum(idxs))
+        The data
+
+    idxs: np.array-like,(sum(p),)
+        The observation locations
+
+    weights : np.array-like,(N,),default=None
```

Comment:
could we specify what we expect in terms of weights here?  if the user passes np.ones(N) vs np.ones(N) / N they'll get different answers

Reply:
Non negative and made those examples equivalent.